k2/csrc/fsa_algo.cu

```diff
@@ -1991,4 +1991,323 @@ FsaOrVec RemoveEpsilonSelfLoops(FsaOrVec &src,
+     }
+     num_arcs_for_states_data[idx01] = num_arcs;
+  });
+
+  ExclusiveSum(num_arcs_for_states, &num_arcs_for_states);
+  RaggedShape states_to_arcs_shape = RaggedShape2(
+      &num_arcs_for_states, nullptr, -1);
+
+  RaggedShape arcs_shape = ComposeRaggedShapes(
+      seqs_to_states_shape, states_to_arcs_shape);
+  int32_t num_arcs = arcs_shape.NumElements();
+
+  // Each state (before merging) has a leaving arc, we add a final arc
+  // to each sequence, so, the total number of arcs equals to
+  // num_states + num_seqs
+  K2_CHECK_EQ(num_arcs, num_seqs + num_states);
+
+  // Populate arcs.
+  // "ss" is short for "sorted states"
+  const int32_t *sampled_paths_data = sampled_paths.values.Data(),
+                *arcs_shape_row_ids1_data = arcs_shape.RowIds(1).Data(),
+                *arcs_shape_row_splits1_data = arcs_shape.RowSplits(1).Data(),
+                *arcs_shape_row_ids2_data = arcs_shape.RowIds(2).Data(),
+                *states_row_ids2_data = states.RowIds(2).Data(),
+                *ss_row_ids1_data = sorted_states.RowIds(1).Data();
+  const float *sampling_probs_data = sampling_probs.values.Data();
+  Array1<Arc> arcs(c, num_arcs);
+  Arc *arcs_data = arcs.Data();
+
+  // The arc_map mapping from lattice arcs to original state indexes.
+  Array1<int32_t> raw_arc_map(c, num_arcs);
+  int32_t *raw_arc_map_data = raw_arc_map.Data();
+
+  K2_EVAL(
+      c, num_arcs, lambda_set_arcs, (int32_t idx012) -> void {
+      Arc arc;
+      int32_t arc_map_value = -1;
+      int32_t idx01 = arcs_shape_row_ids2_data[idx012],
+              idx0 = arcs_shape_row_ids1_data[idx01],
+              idx0x = arcs_shape_row_splits1_data[idx0],
+              idx1 = idx01 - idx0x;
+      arc.src_state = idx1;
+
+      // Final arc of the last sequence.
+      if (idx012 == num_arcs - 1) {
+        arc.dest_state = idx1 + 1;
+        arc.label = -1;
+        arc.score = 0.0;
+      } else {
+        int32_t idx01_next = arcs_shape_row_ids2_data[idx012 + 1],
+                idx0_next = arcs_shape_row_ids1_data[idx01_next];
+        // Final arc for each sequence, except the last sequence.
+        if (idx0 != idx0_next) {
+          arc.dest_state = idx1 + 1;
+          arc.label = -1;
+          arc.score = 0.0;
+        } else {
+          // ss_idx01 is the global index of sorted states, minus idx0 here
+          // because we added an extra final arc for each sequence.
+          int32_t ss_idx01 = idx012 - idx0,
+                  states_idx012 = sorted_states_new2old_data[ss_idx01];
+
+          arc_map_value = states_idx012;
+          arc.label = sampled_paths_data[states_idx012];
+          float sampling_prob = sampling_probs_data[states_idx012];
+
+          int32_t us_idx0 = us_row_ids1_data[ss_idx01],
+                  repeat_num = us_row_splits1_data[us_idx0 + 1] -
+                    us_row_splits1_data[us_idx0];
+
+          arc.score = -logf(1 - powf(1 - sampling_prob, repeat_num));
```

Comment:
I only include the "predictor" head output in C++ part, the other two scores (i.e. hybrid output and lm_output) will add on python part, it would be easier to enable autograd for hybrid output.